src/malloc/wasm32/expand_heap.c

```diff
@@ -0,0 +1,29 @@
+#include <limits.h>
+#include <stdint.h>
+#include <errno.h>
+#include <sys/mman.h>
+#include "libc.h"
+#include "syscall.h"
+
+#define WASM_PAGE_SIZE 65536
+
+/* Expand the heap in-place use WebAssembly grow_memory operators.
+ * The caller is responsible for locking to prevent concurrent calls. */
+
+void *__expand_heap(size_t *pn)
+{
+	size_t n = *pn;
+	n += -n & WASM_PAGE_SIZE-1;
+	unsigned delta = n / WASM_PAGE_SIZE;
+
+	unsigned res;
+	asm("get_local %1; grow_memory 0; set_local %0": "=r"(res) : "r"(delta));
```

Comment:
Cool, does this actually work?

Presumably we could use an intrinsic here too, but I don't know if we have one?  @dschuff ?

Reply:
I'm getting stuck on this line when trying to build:

```text
clang -std=c99 -nostdinc -ffreestanding -Wa,--noexecstack -D_XOPEN_SOURCE=700 -I/src/musl/arch/wasm32 -I/src/musl/arch/generic -Iobj/src/internal -I/src/musl/src/internal -Iobj/include -I/src/musl/include  -g -pipe -fno-unwind-tables -fno-asynchronous-unwind-tables -ffunction-sections -fdata-sections -Werror=implicit-function-declaration -Werror=implicit-int -Werror=pointer-sign -Werror=pointer-arith -Qunused-arguments --target=wasm32-unknown-unknown-wasm -O3 -c -o obj/src/math/__expo2f.o /src/musl/src/math/__expo2f.c
/src/musl/src/malloc/wasm32/expand_heap.c:20:2: error: implicit declaration of function 'asm' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        asm("get_local %1; grow_memory 0; set_local %0": "=r"(res) : "r"(delta));
        ^
/src/musl/src/malloc/wasm32/expand_heap.c:20:49: error: expected ')'
        asm("get_local %1; grow_memory 0; set_local %0": "=r"(res) : "r"(delta));
                                                       ^
/src/musl/src/malloc/wasm32/expand_heap.c:20:5: note: to match this '('
        asm("get_local %1; grow_memory 0; set_local %0": "=r"(res) : "r"(delta));
           ^
2 errors generated.
```

Here is the Dockerfile I'm using to build this:

https://github.com/LinusU/js-webp/blob/8f7968f1acc3101defa668e871c04c99f35e4a7c/Dockerfile#L72

I'm probably missing something simple though, I haven't had much experience with clang -> wasm before ☺️

Reply:
My experiments can be found at https://github.com/yurydelendik/wasmception/

Reply:
Yeah, I think that I'm running the same steps as you are doing. But I'm using a prebuilt LLVM from their official APT-repo instead of building it from source...

Reply:
@yurydelendik, maybe it is better to use __builtin_wasm_memory_size/__builtin_wasm_memory_grow intristics instead of raw asm block?

Reply:
`__builtin_wasm_memory_grow` allowed me to compile that file, had to fix a few other errors as well, but then I could continue:

https://github.com/jfbastien/musl/compare/wasm-prototype-1...LinusU:memory-grow

Now I'm stuck on getting a `libclang_rt.builtins-wasm32.a` file built though ☺️ 😫

Reply:
Compiler-rt should built without any changes on wasm.   We build it along with this copy of musl pas part of the wasm waterfall:
https://wasm-stat.us/console
Src: https://github.com/WebAssembly/waterfall
e.g: https://wasm-stat.us/builders/linux/builds/36342/steps/compiler-rt

Reply:
Hmm, it's probably something with my environment. The error I'm getting is:

```text
In file included from /src/compiler-rt/lib/builtins/comparetf2.c:41:
In file included from /src/compiler-rt/lib/builtins/fp_lib.h:24:
In file included from /usr/lib/llvm-8/lib/clang/8.0.0/include/stdint.h:61:
/usr/include/stdint.h:26:10: fatal error: 'bits/libc-header-start.h' file not found
#include <bits/libc-header-start.h>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

Here is the `Dockerfile` I'm using, my goal is to get an environment for compiling simple C-code into lean webassembly using Docker.

https://github.com/LinusU/js-webp/blob/1273ee38eb59f1b774152a681cad2768a92cc9fc/Dockerfile#L52-L64

Reply:
Ahhh, of course, it's so obvious now that you say it 🤦‍♂️ 

Thank you so much! I'll try this out later today ❤️

Reply:
Hmmmm, actually, it seems like `malloc` isn't working for me 🤔 

```text
/Users/linus/coding/js-webp/index.js:14
  __syscall6 (a, b, c, d, e, f) { throw new Error('Not implemented') },
                                  ^

Error: Not implemented
    at __syscall6 (/Users/linus/coding/js-webp/index.js:14:41)
    at __mmap (wasm-function[2031]:96)
    at malloc (wasm-function[2002]:132)
    at decode (wasm-function[1585]:178)
    at Object.exports.decode (/Users/linus/coding/js-webp/index.js:38:44)
    at Object.<anonymous> (/Users/linus/coding/js-webp/test.js:5:21)
    at Module._compile (internal/modules/cjs/loader.js:689:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:700:10)
    at Module.load (internal/modules/cjs/loader.js:599:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:538:12)
```

For some reason, `malloc` results in a syscall, which I guess _shouldn't_ be the case with this patch, right?

Also, I have to append `/src/compiler-rt/lib/builtins/*.c` to the list of files passed to clang, which doesn't seem right 🤔 

https://github.com/LinusU/js-webp/blob/519115c65e334b6d5c2dc5317dc94f31f49023c5/Dockerfile#L75

Reply:
malloc from dlmalloc/ptmalloc/ptmalloc2/ptmalloc3 allocators uses two syscall types brk/sbrk and mmap. The first one used for increasing size of so called break area (to increase empty top chunk size). The second one serves malloc requests with size bigger that [MMAP_THRESHOLD](https://github.com/jfbastien/musl/blob/0ed10f0b97192e9798fbb23dcd284cb5fe4df47a/src/internal/malloc_impl.h#L23). If you are interested in details please find it [here](https://sploitfun.wordpress.com/2015/02/10/understanding-glibc-malloc/).

Reply:
> For some reason, `malloc` results in a syscall, which I guess _shouldn't_ be the case with this patch, right?

Strike that, apparently, that is supposed to happen, one needs to implement `mmap2` to actually grow the memory. Easily done 😁 

example: https://github.com/yurydelendik/wasmception/commit/671c462eca99c4accce7abd9d22ed5cbb44c42b4

Now I have a new problem 😁 (of course 😅): for some reason `free` is calling `abort` 🤔 This could possibly because I'm doing something wrong with my usage of the webp library though 🤔 

```
/Users/linus/coding/js-webp/index.js:55
    throw new Error(`Syscall ${n} not implemented`)
    ^

Error: Syscall 175 not implemented
    at syscall (/Users/linus/coding/js-webp/index.js:55:11)
    at __syscall4 (/Users/linus/coding/js-webp/index.js:14:39)
    at __block_app_sigs (wasm-function[2036]:16)
    at raise (wasm-function[2038]:23)
    at abort (wasm-function[1998]:3)
    at free (wasm-function[2011]:81)
    at WebPSafeFree (wasm-function[1578]:84)
    at CustomTeardown (wasm-function[98]:78)
    at VP8ExitCritical (wasm-function[48]:224)
    at VP8Decode (wasm-function[160]:500)
```

-----

I still have two strange things related to LLVM though, 1) I have to include `compiler-rt/lib/builtins/*.c` in the list of files I pass to clang when building my project and 2) I have to pass `--export-all`, otherwise no functions gets exported, not even those I've marked with `__attribute__((visibility("default")))` 🤔

Reply:
> malloc from dlmalloc/ptmalloc/ptmalloc2/ptmalloc3 allocators uses two syscall types brk/sbrk and mmap
> So __expand_heap eliminates only brk/sbrk syscalls.

This patch eliminates both (plus futex call) -- you will not see any syscall imports with simple C program that uses malloc.

>  one needs to implement mmap2

This patch takes care of it

> Now I have a new problem ...

I'm not sure how your comments above related to this pull request. Can you try using sysroot created by waterfall (or, at least, try one from wasmception?)

Reply:
@yurydelendik It seems that you are speaking about simple_malloc (function for static
linking in programs which do not call realloc or free)? I tried compiling a program with your sysroot and only one simple malloc - in this case, [simple_malloc] (https://github.com/jfbastien/musl/blob/918b1c1d177b5e3cf22a8aae4a01776495fdc3bc/src/malloc/lite_malloc.c#L11) has been used. But then I just added free call, and standard dlmalloc with these imports:
>   (import "env" "__syscall0" (func $__syscall0 (type 0)))
>   (import "env" "__syscall3" (func $__syscall3 (type 1)))
>   (import "env" "__syscall6" (func $__syscall6 (type 2)))
>   (import "env" "__syscall2" (func $__syscall2 (type 3)))
>   (import "env" "__syscall4" (func $__syscall4 (type 4)))

has appeared.

Reply:
> It seems that you are talking about simple_malloc 

Yeah, that's what clang replaced my malloc calls with. Let's try to remove these as well

Reply:
Sounds good. Does anyone have any thoughts on this?

Reply:
> I'm not sure how your comments above related to this pull request

True, I got a bit carried away 🙈 was so happy to get some pointers on how to do this!

I'm using a sysroot built with this pull request landed, and I just realised that some smaller `malloc` calls were working. So as already noted now, `mmap2` is only needed in JS land when a larger chunk of memory is allocated.

> try one from wasmception

My setup is very similar to wasmception, in fact I'm trying to match your Makefile very closely, just in a Dockerfile. I had problems with building llvm from source though so opted to use their nightly apt builds instead...

Would love to see the other usages of malloc also implemented without syscall. Would be happy to help out in any way I can! If anyone could outline some simple steps of the general direction to take, I would be happy to try and take a stab at it ☺️ 

That being said, it would also be nice to merge this PR since it covers one of the cases. I think that covering more cases could be done in follow up PRs...

Reply:
This seems to be the relevant part where it calls out to `mmap`:

https://github.com/jfbastien/musl/blob/0ed10f0b97192e9798fbb23dcd284cb5fe4df47a/src/malloc/malloc.c#L293-L302

Maybe we could provide a wasm-implementation of `mmap`, that calls `__builtin_wasm_memory_grow` when called with `MAP_ANONYMOUS` (and maybe calls out to a specific JS function to load a file into memory on `MAP_FILE`?)

Reply:
Success 🎉 

The following `mmap` implementation allows me to skip the `mmap` syscall in JS land: https://github.com/jfbastien/musl/commit/a8f52653425c135b798d3537088df335a7917e35

Reply:
I also added a no-op `madvise`, since that is used during `free`: https://github.com/jfbastien/musl/commit/e91c1c93e5a9136f761507dca280573aa0b3f6c6

Reply:
> Sounds good. Does anyone have any thoughts on this?

Yep, pushed more stubs to weed out mmap/madvise/signals syscalls. (Also published complete example with malloc/free exports at https://github.com/yurydelendik/old-man-sandbox/tree/master/nosyscall)

Reply:
@LinusU, there are two possible variants in our case:
1. simple_malloc realization without free operation. In that case this mmap realization has a right to life. But it is not so simple - i will describe it in the next message.
2. dlmalloc realization with full optimization and freeing "in the box". For that case this mmap realization is not so good... To cut a long story short in a "traditional" linux the heap is placed on a so-called break area that can grow (by brk/sbr syscalls) and shrink (by madvice) only from one side. On the other hand there is also mmap syscall that allocates free memory somewhere in the process address space. It means that mmap can't allocate memory from the break area and the "mmap" memory chunks aren't intersected with the "break" chunks. 
Let's consider what will be in this mmap realization. Suppose we have some memory state S = [0, current_brk_area) where the current_brk_area is some constant from dlmalloc inner realization (in fact it is usually used not as a separate constant but the top chunk size from it's header). Then imagine the situation when there is no free memory in our S and a user tries to allocate a big memory chunk so dlmalloc will use mmap for it. And by this realization memory.grow expands Wasm memory by the supplied value. Then, note that  now current_brk_area won't be corresponding to the real Wasm memory limit (call it expanded_memory_area)! Let's imagine the situation when a user tries to allocate small amount of memory from S that still doesn't have enough free memory - so again memory.grow will be called. But it expands not S but [0, expanded_memory_area) to some new area [0, expanded_current_brk_area). So it is possible to allocate chunks from [current_brk_area, expanded_memory_area] which has already allocated another chunks from previous mmap call.

Reply:
@yurydelendik I thougth about it but in this case we have a perfomance issue: all chunks that size is bigger then MMAP_THRESHOLD (0xE000 bytes) are located in the one [bin](https://github.com/jfbastien/musl/blob/wasm-prototype-1/src/malloc/malloc.c#L92). It may also cause a big memory fragmentation. But it seems that there is no simpler way to implement this and in future it is possible to tune this malloc realization.